Clean up Labelers, add test cases, add new Labelers

Miking98 opened the pull request:
* Adds test cases for most labelers
* Adds EHR-ML labelers for common tasks (inpatient mortality, long LOS, readmission)
* Adds lab-based labelers (from Lawrence)
* Adds placeholder labeler to generate labels on admission/discharge times for reusable featurizations
* Bug fixes

Note: `OpioidOverdoseLabeler` and `CeliacTestLabeler` is still a TODO

Lalaland reviewed (changes requested):
Really nicely done! I like the way the unit tests are structured and the generic classes that make it easier to build labelers.

Just some minor comments.

Miking98 commented on src/piton/__init__.py:
```diff
@@ -64,6 +64,7 @@ def __getattr__(self, __name: str) -> Any:
     def __setattr__(self, name: str, value: Any) -> None:
         if value is not None:
             self.__dict__[name] = value
+        self.__dict__[name] = value
```
Oh sorry that was an error when I merged -- will remove!

Miking98 commented on tests/labelers/test_CodeLabelers.py:
```diff
@@ -0,0 +1,510 @@
+
+
+# Local testing
+if __name__ == "__main__":
```
Sometimes its just easier to run `python3 tests/test_specific_test.py` than to deal with all the extra junk that `pytest` logs to the console

Miking98 commented on src/piton/labelers/omop_lab_values.py:
```diff
@@ -0,0 +1,435 @@
+    TimeHorizonEventLabeler,
+)
+from .omop import (
+    _get_all_children,
```
Added a note to this effect!

Miking98 commented on src/piton/labelers/core.py:
```diff
@@ -444,66 +469,78 @@ def label(self, patient: Patient) -> List[Label]:
-        # # checks that we have a label for each prediction time (even if `None``)
-        # assert len(results) == len(self.get_prediction_times(patient))
+        # checks that we have a label for each prediction time (even if `None``)
+        assert len(results) == len(self.get_prediction_times(patient))
```
good call yep

Miking98 commented on src/piton/labelers/omop.py:
```diff
@@ -0,0 +1,690 @@
+    (stored in `omop_concept_ids`) to Piton codes (stored in `codes`)."""
+
+    # names for OMOP concept IDs
+    original_omop_concept_codes: List[str] = []
```
Yeah I removed the other OMOP concept stuff we discussed earlier today re: replicating lawrence's results

Miking98 commented on src/piton/labelers/core.py:
```diff
@@ -444,66 +469,78 @@ def label(self, patient: Patient) -> List[Label]:
             while (
-                curr_outcome_idx < len(outcome_times)
-                and outcome_times[curr_outcome_idx] < time + time_horizon.start
+                curr_outcome_idx < len(outcome_times) - 1
```
It's not necessary with the `curr_outcome_idx < len(outcome_times)` check later in the code, so will remove it

Miking98 commented on src/piton/labelers/omop_lab_values.py:
```diff
@@ -0,0 +1,435 @@
+        admission_events: List[Event] = get_inpatient_admission_events(
+            patient, self.ontology
+        )
+        return [x.start for x in admission_events]
```
What is this?

Lalaland replied:
Resolving as discussed in another comment.

Miking98 commented on src/piton/labelers/core.py:
```diff
@@ -351,17 +254,127 @@ def __len__(self):
+            assert patients is not None
+            # Use `patients` if specified
+            num_patients = len(patients) if not num_patients else num_patients
+            pids = [p.patient_id for p in patients]
```
Yes, it simply takes the first `num_patients` patients from `patients`  - will add a comment to that effect

Miking98 commented on src/piton/labelers/core.py:
```diff
@@ -444,66 +469,78 @@ def label(self, patient: Patient) -> List[Label]:
                 results.append(Label(time=time, value=False))
+            else:
+                # Is censored
+                results.append(Label(time=time, value=None))
```
Yes, it is adjusted for that!

Miking98 commented on src/piton/labelers/core.py:
```diff
@@ -444,66 +469,78 @@ def label(self, patient: Patient) -> List[Label]:
+                    curr_outcome_idx
+                    < len(outcome_times)
+                )
+                and (
```
It's needed in case there are 0 outcomes and 1+ prediction times -- 

We start with

`curr_outcome_idx = 0`

then

`while (curr_outcome_idx < len(outcome_times) and outcome_times[curr_outcome_idx] < time + time_horizon_start):`

will be false since `len(outcome_times) == 0`

so then if you try to do `outcome_times[curr_outcome_idx]` on line 519, you will get an error since `outcome_times[0]` does not exist

Miking98 commented on src/piton/labelers/omop_inpatient_admissions.py:
```diff
@@ -0,0 +1,196 @@
+)
+
+
+class AdmissionDischargePlaceholderLabeler(Labeler):
```
Moved to helper function outside

Miking98 commented on src/piton/labelers/omop_inpatient_admissions.py:
```diff
@@ -0,0 +1,196 @@
+        )
+        times: List[datetime.datetime] = []
+        for x in admission_events:
+            if x.end is None:
```
done

Miking98 commented on src/piton/labelers/omop_inpatient_admissions.py:
```diff
@@ -0,0 +1,196 @@
+    def __init__(
+        self,
+        ontology: extension_datasets.Ontology,
+        long_time: datetime.timedelta,
```
done

Miking98 commented on src/piton/labelers/omop_inpatient_admissions.py:
```diff
@@ -0,0 +1,196 @@
+        super().__init__(ontology=ontology, time_horizon=time_horizon)
+
+
+class _1WeekLongLOSLabeler(InpatientLongAdmissionLabeler):
```
done

Miking98 commented on src/piton/labelers/omop.py:
```diff
@@ -0,0 +1,690 @@
+        )
+
+
+class HighHbA1cCodeLabeler(Labeler):
```
can add as issue

Miking98 commented on src/piton/labelers/core.py:
```diff
@@ -521,5 +558,4 @@ def label(self, patient: Patient) -> List[Label]:
         return n_labels
```
done

Miking98 commented on src/piton/labelers/omop_inpatient_admissions.py:
```diff
@@ -0,0 +1,196 @@
+                is_long_admission: bool = (
+                    event.end - event.start
+                ) >= self.long_time
+                labels.append(Label(event.start, is_long_admission))
```
Why is this? (same as comment on `omop_lab_values.py`)

rthapa84 replied:
^^

Lalaland replied:
We generally only want to make predictions at 11:59 to capture enough starting information about the admission before we make a prediction.

rthapa84 commented on src/piton/labelers/core.py:
```diff
@@ -351,17 +268,142 @@ def __len__(self):
+        # Split patient IDs across parallelized processes
+        pids_parts = np.array_split(pids, num_threads)
+
+        # NOTE: Super hacky workaround to pickling limitations
```
I am not sure I understand this workaround. Why do we need this hack?

Miking98 replied:
If an object has a `self.ontology` property (which has type `extensions.Ontology`), we won't be able to pickle that object (i.e. `self`) since `extensions.Ontology` is not pickleable. Thus, we won't be able to pass the labeler (i.e. `self`) to multiprocessing b/c it's not pickable

rthapa84 commented on src/piton/labelers/omop_inpatient_admissions.py:
```diff
@@ -0,0 +1,161 @@
+)
+
+
+class DummyAdmissionDischargeLabeler(Labeler):
```
Do we need to call this Dummy?

Miking98 replied:
Ethan suggested that as a name to make it explicit its not a "true" labeler @Lalaland

rthapa84 commented on src/piton/labelers/omop_inpatient_admissions.py:
```diff
@@ -0,0 +1,161 @@
+
+    def get_outcome_times(self, patient: Patient) -> List[datetime.datetime]:
+        """Return the start times of inpatient admissions."""
+        admission_events: List[Event] = get_inpatient_admission_events(
```
Here, don't we need to skip the first admission_events? The outcome to this task can also be seen after the first visit, no?

Miking98 commented on src/piton/labelers/omop.py:
```diff
@@ -0,0 +1,690 @@
+##########################################################
+
+
+class WithinVisitLabeler(Labeler):
```
Added a `WithinInpatientVisitLabeler` class

Miking98 commented:
* Updated lab value labelers to only consider visits where a lab result was returned
* Cleaned up admission definitions combining best of visit + visit detail tables

Lalaland commented on README.md:
````diff
@@ -34,7 +34,7 @@ export DISTDIR=/local-scratch/nigam/distdir
 
 ```
-conda create -n PITON_ENV python=3.10 bazel=5.3 clangxx=14 -c conda-forge
+conda create -n PITON_ENV python=3.10 bazel=5.3 clangxx=14 pathos=0.3.0 -c conda-forge
````
What is pathos for?

Miking98 replied:
**pathos** is just a more robust version of **multiprocessing** that includes **dill** as its pickle module by default. 

**dill** is just a more robust version of **pickle** that allows us to pickle classes (in this case, Labelers) that have more "exotic" contents, e.g. arguments that are lambdas, certain properties within the class that aren't naturally pickelable, etc.

It's a super minimal change code-wise -- literally just change `multiprocessing.get_context("forkserver").Pool(num_threads)` to `pathos.ProcessPool`

Miking98 replied:
I added it to the `pyproject.toml` as well, so I can remove it from this command line command if that's redundant (I'm still not quite clear conceptually how the dependencies in `pyproject.toml` relate to these conda dependencies / how they're installed?)

Miking98 replied:
@Lalaland

Lalaland replied:
Ah, that sounds pretty handy and good. Yeah, remove it from here and put in pyproject.toml.

The dependencies here are only here because they cannot be in pyproject.toml (due to not being Python packages).

Lalaland reviewed (approved):
Looks good to me! Just one question about the pathos library.

Miking98 commented:
Hey sorry for all the back and forth on this

Basically, I reverted everything back to the old definition of visits -- Anything from the `visit_occurrence` table with `e.code == 'Visit/IP'` and `e.start != e.end` is now regarded as an inpatient visit.

I rewrote all the test cases to reflect this fact, and fixed all the previous changes I'd introduced when messing around with `visit_detail`

Miking98 commented:
I think this should be good now

Miking98 commented:
Hmm actually it looks like Github actions does not like this line in **pyproject.toml:**

```
"git+https://github.com/anivegesana/dill.git@enums",
```

but this patch is needed to run the tests on classes that inherit from abstract classes.

Let me see if I can figure out a workaround.

Miking98 commented:
OK should be fixed now -- I added an issue to resolve this in the future

Lalaland commented:
Looks good to me!
